From ab33d7524f04aaaf78fd3d5dd5bfe905f2d31386 Mon Sep 17 00:00:00 2001
From: qwang2 <quanyang.wang@windriver.com>
Date: Sun, 18 Nov 2018 19:17:43 +0800
Subject: [PATCH] nxp-imx8: use dev_pm_opp_put to put reference

In upstream commit 8a31d9d94297 ("PM / OPP: Update OPP users to put reference"),
rcu_read_lock/unlock has been moved inside the function "dev_pm_opp_find_freq_ceil",
and in upstream commit 052c6f19141d ("PM / OPP: Move away from RCU locking"),
rcu_read_lock/unlock has been change to mutex_lock/unlock.

So we should delete rcu_read_lock/unlock which outsides the dev_pm_opp_find_freq_ceil,
and use dev_pm_opp_put to put reference.

This patch fix the calltrace as below:

BUG: sleeping function called from invalid context at kernel/locking/mutex.c:747
in_atomic(): 0, irqs_disabled(): 0, pid: 37, name: kworker/2:1
INFO: lockdep is turned off.
Preemption disabled at:
[<ffff000008bb5b28>] __mutex_lock+0x54/0x878
CPU: 2 PID: 37 Comm: kworker/2:1 Tainted: G        W       4.12.29-yocto-standard #276
Hardware name: Freescale i.MX8MQ EVK (DT)
Workqueue: events dbs_work_handler
Call trace:
[<ffff00000808c518>] dump_backtrace+0x0/0x24c
[<ffff00000808c788>] show_stack+0x24/0x30
[<ffff0000085685c4>] dump_stack+0xac/0xe4
[<ffff0000080da664>] ___might_sleep+0x154/0x228
[<ffff0000080da790>] __might_sleep+0x58/0x8c
[<ffff000008bb5b20>] __mutex_lock+0x4c/0x878
[<ffff000008bb6388>] mutex_lock_nested+0x3c/0x4c
[<ffff00000871292c>] _find_opp_table+0x3c/0x88
[<ffff000008712e38>] dev_pm_opp_find_freq_ceil+0x30/0x98
[<ffff0000088955f8>] imx8mq_set_target+0xc0/0x358
[<ffff00000888a774>] __cpufreq_driver_target+0x274/0x5b8
[<ffff00000888f3ac>] od_dbs_update+0xac/0x1ac
[<ffff000008893988>] dbs_work_handler+0x48/0x80
[<ffff0000080c9c84>] process_one_work+0x1f0/0x6c4
[<ffff0000080ca2a4>] worker_thread+0x14c/0x40c
[<ffff0000080d13fc>] kthread+0x114/0x140
[<ffff000008086620>] ret_from_fork+0x10/0x20

Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/cpufreq/imx8mq-cpufreq.c |    4 +---
 1 file changed, 1 insertion(+), 3 deletions(-)

diff --git a/drivers/cpufreq/imx8mq-cpufreq.c b/drivers/cpufreq/imx8mq-cpufreq.c
index ee3725d..b0b4446 100644
--- a/drivers/cpufreq/imx8mq-cpufreq.c
+++ b/drivers/cpufreq/imx8mq-cpufreq.c
@@ -48,15 +48,13 @@ static int imx8mq_set_target(struct cpufreq_policy *policy, unsigned int index)
 	freq_hz = new_freq * 1000;
 	old_freq = policy->cur;
 
-	rcu_read_lock();
 	opp = dev_pm_opp_find_freq_ceil(cpu_dev, &freq_hz);
 	if (IS_ERR(opp)) {
-		rcu_read_unlock();
 		dev_err(cpu_dev, "failed to find OPP for %ld\n", freq_hz);
 		mutex_unlock(&set_cpufreq_lock);
 		return PTR_ERR(opp);
 	}
-	rcu_read_unlock();
+	dev_pm_opp_put(opp);
 
 	dev_dbg(cpu_dev, "%u MHz --> %u MHz\n",
 		old_freq / 1000, new_freq / 1000);
-- 
1.7.9.5

